src/Mod/Sketcher/Gui/SnapManager.cpp

```diff
@@ -0,0 +1,309 @@
+        auto function = val.second;
+
+        function(string);
+    }
+}
+
+void SnapManager::ParameterObserver::updateSnapToObjectParameter(const std::string& parametername)
+{
+    (void)parametername;
+
+    ParameterGrp::handle hGrp = App::GetApplication().GetParameterGroupByPath("User parameter:BaseApp/Preferences/Mod/Sketcher/General");
+
+    Client.snapToObjects = hGrp->GetBool("SnapToObjects", true);
+}
+
+void SnapManager::ParameterObserver::updateSnapAngleParameters(const std::string& parametername)
+{
+    (void)parametername;
+
+    ParameterGrp::handle hGrp = App::GetApplication().GetParameterGroupByPath("User parameter:BaseApp/Preferences/Mod/Sketcher/General");
+
+    Client.snapAngle = hGrp->GetFloat("SnapAngle", 5.) * M_PI / 180;
+}
+
+
+void SnapManager::ParameterObserver::subscribeToParameters()
+{
+    try {
+        ParameterGrp::handle hGrp = App::GetApplication().GetParameterGroupByPath("User parameter:BaseApp/Preferences/Mod/Sketcher/General");
+        hGrp->Attach(this);
+    }
+    catch (const Base::ValueError& e) { // ensure that if parameter strings are not well-formed, the exception is not propagated
+        Base::Console().Error("SnapManager: Malformed parameter string: %s\n", e.what());
+    }
+}
+
+void SnapManager::ParameterObserver::unsubscribeToParameters()
+{
+    try {
+        ParameterGrp::handle hGrp = App::GetApplication().GetParameterGroupByPath("User parameter:BaseApp/Preferences/Mod/Sketcher/General");
+        hGrp->Detach(this);
+    }
+    catch (const Base::ValueError& e) {// ensure that if parameter strings are not well-formed, the program is not terminated when calling the noexcept destructor.
+        Base::Console().Error("SnapManager: Malformed parameter string: %s\n", e.what());
+    }
+}
+
+void SnapManager::ParameterObserver::OnChange(Base::Subject<const char*>& rCaller, const char* sReason)
+{
+    (void)rCaller;
+
+    auto key = str2updatefunction.find(sReason);
+    if (key != str2updatefunction.end()) {
+        auto string = key->first;
+        auto function = key->second;
+
+        function(string);
+    }
+}
+
+
+//**************************** SnapManager class ******************************
+
+SnapManager::SnapManager(ViewProviderSketch &vp):viewProvider(vp), angleSnapEnabled(false), angleSnapRef(Base::Vector2d(0.,0.)) {
+    // Create parameter observer and initialise watched parameters
+    pObserver = std::make_unique<SnapManager::ParameterObserver>(*this);
+}
+
+SnapManager::~SnapManager() {}
+
+void SnapManager::snap(double& x, double& y)
+{
+    //In order of priority :
+
+    // 1 - Snap at an angle
+    if (angleSnapEnabled && QApplication::keyboardModifiers() == Qt::ControlModifier) {
+        snapAtAngle(x, y);
+        return;
+    }
+
+    // 2 - Snap to objects
+    if (snapToObjects 
+        && snapToObject(x, y)) {
+        return;
+    }
+    
+    // 3 - Snap to grid
+    if (viewProvider.GridSnap.getValue() && viewProvider.ShowGrid.getValue()) {

```

Comment:
And this is what I was looking for... This is how the property of the grid extension that needs to be moved out affects this PR.

I need to think about what is best here. Will come back.